### PR TITLE
Autotuning support for RISE using Hypermapper

### PR DESCRIPTION
**Pull request by johanneslenfers**
# Autotuning for RISE
Parameter auto tuning using Hypermapper via stdin/stdout in client-server mode 

## Features
- definition of tuning parameters
- generation of Hypermapper config file 
- configuration of tuning run

## Example
 ```scala
 // define main 
 val main: Int => String = iterations => {
    s"""
    const int N = ${iterations};
    int main(int argc, char** argv) {
      Context ctx = createDefaultContext();
      Buffer input = createBuffer(ctx, N * sizeof(float), HOST_READ | HOST_WRITE | DEVICE_READ);
      Buffer output = createBuffer(ctx, N * sizeof(float), HOST_READ | HOST_WRITE | DEVICE_WRITE);

      float* in = hostBufferSync(ctx, input, N * sizeof(float), HOST_WRITE);
      for (int i = 0; i < N; i++) {
        in[i] = 1;
      }

      foo_init_run(ctx, output, input, input);

      float* out = hostBufferSync(ctx, output, N * sizeof(float), HOST_READ);

//    add error checking here 

      destroyBuffer(ctx, input);
      destroyBuffer(ctx, output);
      destroyContext(ctx);
      return EXIT_SUCCESS;
    }
    """
  }

  // define expression with tuning params
  val convolutionOcl: ToBeTyped[Expr] =
  // tileShiftInwards should constrain n >= tile
  // slideVectors and slide should constrain tile % vec = 0
    tuningParam("vec", RangeAdd(0, 32, 1), (vec: Nat) =>
      tuningParam("tile", RangeAdd(4, 32, 1), (tile: Nat) =>
        depFun(RangeAdd(1, PosInf, vec), (n: Nat) =>
          fun(3 `.` f32)(weights =>
            fun(((n + 2) `.` f32) ->: (n `.` f32))(input =>
              oclRun(LocalSize(1), GlobalSize(32))(
                input |> tileShiftInwards(tile)(mapWorkGroup(0)(
                  slideVectors(vec) >> slide(3)(vec) >>
                    mapLocal(0)(weightsSeqVecUnroll(weights)) >>
                    asScalar
                ))
              ))))))


  val e: Expr = convolutionOcl(32)
  val tuningResult = autotune.search(Tuner(main(32)))(e)     // performs auto tuning 

  val bestSample = autotune.getBest(tuningResult.samples)    // get best result 
  println("bestSample: \n" + bestSample)

  autotune.saveSamples("autotuning/RISE.csv", tuningResult)  // save results 
```

## Problems to solve  
- Merge with master fails due to changes e.g. in [src/main/scala/rise/core/types/Nat.scala](https://github.com/rise-lang/shine/blob/master/src/main/scala/rise/core/types/Nat.scala)
- One test is failing, which tests the wrapping of the `oclRun` pattern. [src/test/scala/rise/autotuning.scala](https://github.com/rise-lang/shine/blob/autotuning_merge/src/test/scala/rise/autotuning.scala)

### CODE REVIEW

File: src/main/scala/rise/core/types/Nat.scala [hunk 1]
```diff
@@ -5,41 +5,52 @@ import rise.core.types
 
 class NatIdentifier(
     override val name: String,
-    override val range: Range = RangeUnknown,
-    override val isExplicit: Boolean = false
+    override val range: Range,
+    override val isExplicit: Boolean,
+    val isTuningParam: Boolean
 ) extends NamedVar(name, range)
```
Comment (Bastacyclop) [inline src/main/scala/rise/core/types/Nat.scala:11]: @umazalakain any insight on how to translate this into the new identifier design?

Comment (umazalakain) [reply src/main/scala/rise/core/types/Nat.scala:11]: Would it be possible to keep a set of globally unbound identifier names that are meant to be tuned?

File: src/main/scala/util/package.scala [hunk 2]
```diff
@@ -1,6 +1,24 @@
 import java.io.{File, PrintWriter}
+import java.util.concurrent.{Executors, TimeUnit, TimeoutException}
 
 package object util {
+
+  // WARNING: does not kill the thread, we only return to host program
+  // Thread.stop() is deprecated and cannot be used here
```
Comment (Bastacyclop) [inline src/main/scala/util/package.scala:7]: is this good enough when autotuning?

**Review by Bastacyclop — CHANGES_REQUESTED** [review]

**Comment by Bastacyclop** [comment]
@johanneslenfers Can you explain what the included json files are? Are they written by hand or generated?

File: autotuning/hm_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    """
+
+    # change directory to ~
+    os.chdir(os.path.expanduser("~"))
+
+    cmd = [
+        ".local/bin/hypermapper",
+        str(config_file)
+    ]
+    # print("comanndon")
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply autotuning/hm_wrapper.py]: deleted

File: mmTuning [hunk 4]
```diff
@@ -0,0 +1,3 @@
...
...2.13/2.9.1/play-json_2.13-2.9.1.jar:/home/jo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/jline/jline/3.15.0/jline-3.15.0.jar:/home/jo/.cache/coursier/v1/https/repo1.maven.org/maven2/net/java/dev/jna/jna/5.3.1/jna-5.3.1.jar:/home/jo/.cache/coursier/v1/https/repo1.maven.org/maven2/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar:/home/jo/.cache/coursier/v1/https/repo1.maven.org/maven2/com/typesafe/play/play-functional_2.13/2.9.1/play-functional_2.13-2.9.1.jar:/home/jo/.cache/coursier/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-core/2.10.5/jackson-core-2.10.5.jar:/home/jo/.cache/coursier/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-annotations/2.10.5/jackson-annotations-2.10.5.jar:/home/jo/.cache/coursier/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/datatype/jackson-datatype-jdk8/2.10.5/jackson-datatype-jdk8-2.10.5.jar:/home/jo/.cache/coursier/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/datatype/jackson-datatype-jsr310/2.10.5/jackson-datatype-jsr310-2.10.5.jar:/home/jo/.cache/coursier/v1/https/repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-databind/2.10.5/jackson-databind-2.10.5.jar apps.mmTuning $*
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply mmTuning]: Yes, we should delete this. Currently this is used for the Hypermapper and ATF experiments. But this should not be part of the PR. Also deleted.

File: scripts/buildRunScripts.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+# for each of the mainclasses, build a runscript
+for clazz in mainClasses:
+	finalCommand = command + " " + clazz + " $*\n"
+	fileName = generatedScriptsFolder + clazz.split(".")[-1]
+	f = open(fileName, "w")
+	f.write(finalCommand)
+	f.close()
+	subprocess.call(["chmod", "u+x", fileName])
+	print("Created a script for " + clazz)
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply scripts/buildRunScripts.py]: This is to create the mmTuning script above. We should also delete this.

File: src/main/scala/apps/mm.scala [hunk 6]
```diff
@@ -70,12 +70,22 @@ object mm {
...
+
+    // A(o,n) x B(o,m)
+    // val v5 = 64 // tile-width A
+    // val v7 = 128 // tile-width B
+    // val v8 = 16 // tile-height A,B
+
+    // val v3 = 4 // divides v7
+    // val v4 = 8 // divides v5
+    // val v6 = 128 // divides v8 x v5
+
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/apps/mm.scala]: deleted

File: src/main/scala/apps/mm.scala [hunk 7]
```diff
@@ -137,19 +147,89 @@ object mm {
     ))
   }
 
+  def genMMKernel(v3:Int, v4:Int, v5:Int, v6:Int, v7:Int, v8:Int):ToBeTyped[Expr] = {
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/apps/mm.scala]: This is also for the ATF tuning. Should be removed. Deleted.

File: src/main/scala/util/ExecuteOpenCL.scala [hunk 8]
```diff
@@ -52,7 +54,64 @@ object ExecuteOpenCL {
...
-          .deleteRecursively()
+        .deleteRecursively()
+    }
+  }
+
+  def getRuntimeFromClap(s: String): TimeSpan[Time.ms] = {
+
+    // Todo: check weird error, where the first line is an empty element
+    // avoid this error by removing first line
+
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/util/ExecuteOpenCL.scala:65]: The error/bug is catched, but I did not find the reason for the error. So I think we can keep it as it is and just remove the "todo".

File: src/main/scala/util/ExecuteOpenCL.scala [hunk 9]
```diff
@@ -52,7 +54,64 @@ object ExecuteOpenCL {
...
+      (Some(runtime), AutoTuningError(NO_ERROR, None))
+    } catch {
+      // todo check error codes here
+      case e: Throwable => {
+        println("execution error: " + e)
+
+        (None, AutoTuningError(EXECUTION_ERROR, Some(e.toString)))
+      }
     }
   }
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/util/ExecuteOpenCL.scala:116]: It is basically a function to run the code with clap. Maybe there is a use case, where you just want to run and measure your program without tuning. But as the auto tuning error levels are reported here, this should be part of the autotuning. I agree.

File: src/main/scala/util/package.scala (hunk 2 above)
Comment (johanneslenfers) [reply src/main/scala/util/package.scala:7]: It is okay when the problem of generating tons of kernel-code only occurs a few times during a tuning run. We should definitely find a solution for that.

Comment (Bastacyclop) [reply src/main/scala/util/package.scala:7]: Maybe you can open an issue to track this then?

File: src/main/scala/exploration/runner/CExecutor.scala [hunk 10]
```diff
@@ -441,6 +441,8 @@ int main(int argc, char** argv) {
     // todo: make this configable using json file
     // compile
     s"clang -O2 $src -o $bin -lm -fopenmp" !!
+//    s"clang -O2 $src -o $bin -lm" !!
+//    s"clang $src -o $bin -fopenmp" !!
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/exploration/runner/CExecutor.scala]: deleted

File: src/main/scala/rise/core/types/Nat.scala (hunk 1 above)
Comment (Bastacyclop) [reply src/main/scala/rise/core/types/Nat.scala:11]: Maybe, but there are some challenges:
1. Ideally we would keep a syntax similar to `tuningParam("vec", RangeAdd(0, 32, 1), (vec: Nat) => /* use vec ... */)` to create tuning parameters in an expression
2. Composing two expressions should also compose their set of tuning parameters (it doesn't really sound like a global set anymore)
3. Rewrite rules should be able to introduce tuning parameters using the same syntax

File: src/main/scala/apps/mmTuning.scala [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+      var rowString = ""
+      row.foreach(elem => {
+        rowString += elem.toString + ","
+      })
+      AtString += rowString.dropRight(1) + "\n"
+    })
+
+
+    AtString
+  }
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/apps/mmTuning.scala:60]: This should not be part of the PR. I deleted it. This was old work in progress.

File: src/main/scala/apps/mmTuning.scala [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+
+//        val costfile = new PrintWriter(new FileOutputStream(new File("/home/jo/development/lift/atf/atfc/build/costfile.txt"), false))
+//        costfile.println(Double.MaxValue.toString)
+//        costfile.close()
+
+        sys.exit(1)
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/apps/mmTuning.scala:272]: same here, it is deleted

**Comment by michel-steuwer** [comment]
@johanneslenfers after merging #186 the files `runtime/buffer_zero_copy.c` and `runtime/ocl.c` have moved to the folder `runtime/ocl/`

It seems that this is the reason why the autotuning tests are now failing: https://github.com/rise-lang/shine/runs/2895321354#step:6:7201

Could you please have a look at this and also address the outstanding questions and points from @Bastacyclop

File: src/main/scala/apps/mm.scala [hunk 13]
```diff
@@ -137,19 +147,89 @@ object mm {
...
-          sum += At(k)(i) * B(k)(j)
+      for (k <- 0 until o) {
+          for (j <- 0 until m) {
+          output(i)(j) += At(k)(i) * B(k)(j)
+//          sum += At(k)(i) * B(k)(j)
+//          output(i)(j) += sum
         }
 
-        output(i)(j) = sum
       }
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/apps/mm.scala]: It was faster for my inital experiments on the auto tuning. This should not be part of the PR. I removed the changes

File: src/main/scala/util/ExecuteOpenCL.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (Bastacyclop) [reply src/main/scala/util/ExecuteOpenCL.scala:115]: You did not move the function?

File: src/test/scala/rise/autotuning.scala [hunk 14]
```diff
@@ -0,0 +1,594 @@
...
+                                generate(fun(_ =>
+                                  generate(fun(_ =>
+                                    generate(fun(_ => lf32(0.0f))))))))) |>
+                                mapLocal(1)(mapLocal(0)(mapSeq(mapSeq(id))))
+                            ) |> // (v5/^v4).(v7/^v3).v4.v3.f
+                            mapLocal(1)(mapLocal(0)(mapSeq(asScalar o mapSeq(id) o asVector(4)))) |>
+                            map(transpose) |> join |> map(join) |> transpose // v7.v5.f
+                        )) |> join |> transpose // v5.M.f
+                    )) |> join // N.M.f
+                ))))))))
```
Comment (Bastacyclop) [inline src/test/scala/rise/autotuning.scala:235]: it feels like this same kernel is duplicated at least three times in the code, can we factorize this?

File: src/main/scala/util/package.scala (hunk 2 above)
Comment (Bastacyclop) [reply src/main/scala/util/package.scala:7]: where is the issue?

File: src/main/scala/util/ExecuteOpenCL.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/util/ExecuteOpenCL.scala]: the function is moved

File: src/main/scala/rise/autotune/package.scala [hunk 15]
```diff
@@ -0,0 +1,785 @@
...
+    val output = parameters.filter(nat => {
+      parametersInConstraint.exists(paramInConstraint => {
+        paramInConstraint.name.equals(nat.name)
+      })
+    })
+
+    output
+  }
+
+}
```
[the remark this answers is not in the document]
Comment (johanneslenfers) [reply src/main/scala/rise/autotune/package.scala:260]: I reorganized this file. We now have multiple files implementing the autotuning.

File: src/test/scala/rise/autotuning.scala (hunk 14 above)
Comment (johanneslenfers) [reply src/test/scala/rise/autotuning.scala:235]: Yes, we can. But we use this kernel here only to test the constraints collection and constraints checking for the autotuning. In [mm.scala](https://github.com/rise-lang/shine/blob/autotuning_merge/src/main/scala/apps/mm.scala) we have the non-autotuning version.

File: src/main/scala/util/package.scala (hunk 2 above)
Comment (johanneslenfers) [reply src/main/scala/util/package.scala]: Issue is open

File: src/test/scala/rise/autotuning.scala (hunk 14 above)
Comment (Bastacyclop) [reply src/test/scala/rise/autotuning.scala:234]: You could do something like:
```scala
def mmNVIDIAWithParams(v3: Nat, ...): ToBeTyped[Expr] = ...

val mmKernel = tuningParam("v3", (v3: Nat) => ... mmNVIDIAWithParams(v3, ...))

val mmNVIDIA = mmNVIDIAWithParams(v3 = 4, ...)
```

Comment (Bastacyclop) [reply src/test/scala/rise/autotuning.scala:234]: You could also do:
```
val mmNVIDIA = mmKernel.subsituteNats(Map(v3 -> 4, ...))
```

File: src/main/scala/rise/autotune/autoTuningErrors.scala [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+sealed trait AutoTuningErrorLevel
+
+case object NO_ERROR extends AutoTuningErrorLevel
+case object CONSTRAINTS_ERROR extends AutoTuningErrorLevel
+case object SUBSTITUTION_ERROR extends AutoTuningErrorLevel
+case object CODE_GENERATION_ERROR extends AutoTuningErrorLevel
+case object COMPILATION_ERROR extends AutoTuningErrorLevel
+case object EXECUTION_ERROR extends AutoTuningErrorLevel
+
+case class AutoTuningError(errorLevel: AutoTuningErrorLevel, message:Option[String])
```
Comment (Bastacyclop) [inline src/main/scala/rise/autotune/autoTuningErrors.scala:12]: maybe the file can just be named `autotune/errors.scala`? no need to repeat autotuning. same for the other files.

File: src/main/scala/rise/autotune/package.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (Bastacyclop) [reply src/main/scala/rise/autotune/package.scala:245]: Thanks :+1: , its much easier to navigate

**Review by Bastacyclop — CHANGES_REQUESTED** [review]
Thanks for the additional cleanup, only one final requested change and it looks good to me!

**Comment by johanneslenfers** [comment]
I'm happy to fix the style issues. Seems to be no big deal. I did not know, that we have such a diagnostic tool. Interesting.

File: src/main/scala/rise/autotune/autoTuningErrors.scala (hunk 16 above)
Comment (johanneslenfers) [reply src/main/scala/rise/autotune/autoTuningErrors.scala]: I removed the *autoTuning* prefix from the files.

File: src/test/scala/rise/autotuning.scala (hunk 14 above)
Comment (johanneslenfers) [reply src/test/scala/rise/autotuning.scala]: I created a factorized version. Looks much better having this duplicated 60 lines deleted.

**Review by Bastacyclop — APPROVED** [review]

**Review by Bastacyclop — APPROVED** [review]

**Pull request merged**